waspc/cli/Wasp/Cli/Command/Watch.hs

```diff
@@ -40,15 +43,41 @@ watch waspProjectDir outDir = FSN.withManager $ \mgr -> do
     listenForEvents :: Chan FSN.Event -> UTCTime -> IO ()
     listenForEvents chan lastCompileTime = do
       event <- readChan chan
-      let eventTime = FSN.eventTime event
-      if eventTime < lastCompileTime
-        then -- If event happened before last compilation started, skip it.
+      if isStaleEvent event lastCompileTime
+        then -- Ignore delayed/stale events older than our last compile time.
           listenForEvents chan lastCompileTime
         else do
+          -- Recompile, but only after a 1s period of no new events.
+          waitUntilNoNewEvents chan lastCompileTime 1
           currentTime <- getCurrentTime
           recompile
           listenForEvents chan currentTime
 
+    -- Blocks until no new events are recieved for a duration of `secondsToDelay`.
+    -- Consumes any new events during an active timer window and then restarts wait.
+    -- If a stale event comes in during an active timer window, we immediately
+    -- return control to the caller.
+    waitUntilNoNewEvents :: Chan FSN.Event -> UTCTime -> Int -> IO ()
+    waitUntilNoNewEvents chan lastCompileTime secondsToDelay = do
+      eventOrDelay <- race (readChan chan) (threadDelaySeconds secondsToDelay)
+      case eventOrDelay of
+        Left event -> do
+          if isStaleEvent event lastCompileTime
+            then return ()
+            else -- We have a new event, restart waiting process.
+              waitUntilNoNewEvents chan lastCompileTime secondsToDelay
```

Comment:
You can do
```suggestion
          unless (isStaleEvent event lastCompileTime) $
            -- We have a new event, restart waiting process.
            waitUntilNoNewEvents chan lastCompileTime secondsToDelay
```

There is also an opposite function, `when`.

Reply:
Ah cool, thanks for bringing this up! I was actually originally considering using `unless` but I admittedly wasn't exactly sure how it handled the alternative case from the docs. So when the `unless` fails in an `IO ()` case it defaults to `return ()`? That makes sense. Pretty handy! This does look cleaner! Thnx

```
      case eventOrDelay of
        Left event -> do
          unless (isStaleEvent event lastCompileTime) $
            -- We have a new event, restart waiting process.
            waitUntilNoNewEvents chan lastCompileTime secondsToDelay
        Right () -> return ()
```

Reply:
Yes, basically it does exactly what your previous code was doing, calls return ()! And same goes for `when`.